concent_api/core/views.py

```diff
@@ -1253,6 +1207,450 @@ def store_message_and_message_status(
+    requestor_public_key:           bytes,
+    state:                          Enum,
+    next_deadline:                  int,
+    task_to_compute:                message.TaskToCompute                = None,
+    report_computed_task:           message.ReportComputedTask           = None,
+    ack_report_computed_task:       message.AckReportComputedTask        = None,
+    reject_report_computed_task:    message.RejectReportComputedTask     = None,
+    subtask_results_accepted:       message.tasks.SubtaskResultsAccepted = None,
+    subtask_results_rejected:       message.tasks.SubtaskResultsRejected = None,
+):
+    assert isinstance(task_id,              str)
+    assert isinstance(subtask_id,           str)
+    assert isinstance(provider_public_key,  bytes)
+    assert isinstance(requestor_public_key, bytes)
+    assert state in Subtask.SubtaskState
+
+    if state in Subtask.SUBTASK_ACTIVE_STATES:
+        assert next_deadline is not None
+    elif state in Subtask.SUBTASK_PASSIVE_STATES:
+        assert next_deadline is None
+
+    provider  = Client.objects.get_or_create_full_clean(provider_public_key)
+    requestor = Client.objects.get_or_create_full_clean(requestor_public_key)
+
+    subtask = Subtask(
+        task_id         = task_id,
+        subtask_id      = subtask_id,
+        provider        = provider,
+        requestor       = requestor,
+        state           = state.name,
+        next_deadline   = datetime.datetime.fromtimestamp(next_deadline, timezone.utc),
+    )
+
+    add_messages_to_subtask(
+        subtask,
+        task_to_compute,
+        report_computed_task,
+        ack_report_computed_task,
+        reject_report_computed_task,
+        subtask_results_accepted,
+        subtask_results_rejected,
+    )
+
+    subtask.full_clean()
+    subtask.save()
+    return subtask
+
+
+def update_concent_states(
+    client_public_key: bytes,
+):
+    clients_subtask_list = Subtask.objects.filter(
+        Q(requestor__public_key = b64encode(client_public_key)) | Q(provider__public_key = b64encode(client_public_key)),
+        state__in               = Subtask.SUBTASK_ACTIVE_STATES,
+        next_deadline__lte      = timezone.now()
+    )
+
+    if clients_subtask_list.exists():
+        update_states_of_subtasks(
+            clients_subtask_list,
+            client_public_key,
+        )
+        logging.log_changes_in_subtask_states(
+            b64encode(client_public_key),
+            clients_subtask_list.count(),
+        )
+    else:
+        logging.log_no_changes_in_subtask_states(
+            b64encode(client_public_key)
+        )
+
+
+def update_states_of_subtasks(
+    subtask_list:       list,
+    client_public_key:  bytes,
+):
+
+    for subtask in subtask_list:
+        if subtask.state == Subtask.SubtaskState.FORCING_REPORT:
+            subtask = update_subtask(
+                subtask_id              = subtask.subtask_id,
+                requestor_public_key    = client_public_key,
```

Comment:
Please review this part. Is it always requestor on every case ? Also, in each `store_message_inbox` you use this key as `client_public_key`. Do we always store this message for requestor ?